File: verify-pipeline.mjs
#!/usr/bin/env node
/**
 * verify-pipeline.mjs — Health check for career-ops pipeline integrity
 *
 * Checks:
 * 1. All statuses are canonical (per states.yml)
 * 2. No duplicate company+role entries
 * 3. All report links point to existing files
 * 4. Scores match format X.XX/5 or N/A or DUP
 * 5. All rows have proper pipe-delimited format
 * 6. No pending TSVs in tracker-additions/ (only in merged/ or archived/)
 * 7. states.yml canonical IDs for cross-system consistency
 *
 * Run: node career-ops/verify-pipeline.mjs
 */

import { readFileSync, readdirSync, existsSync, mkdirSync, unlinkSync, statSync } from 'fs';
import { join, dirname } from 'path';
import { fileURLToPath } from 'url';
import { resolveColumns } from './column-map.mjs';

const CAREER_OPS = dirname(fileURLToPath(import.meta.url));
// Support both layouts: data/applications.md (boilerplate) and applications.md (original).
// CAREER_OPS_TRACKER overrides the path (used by tests and non-standard layouts).
const APPS_FILE = process.env.CAREER_OPS_TRACKER
  ? process.env.CAREER_OPS_TRACKER
  : existsSync(join(CAREER_OPS, 'data/applications.md'))
    ? join(CAREER_OPS, 'data/applications.md')
    : join(CAREER_OPS, 'applications.md');
const ADDITIONS_DIR = join(CAREER_OPS, 'batch/tracker-additions');
// CAREER_OPS_REPORTS_DIR overrides the reports directory (used by tests).
const REPORTS_DIR = process.env.CAREER_OPS_REPORTS_DIR || join(CAREER_OPS, 'reports');
const STATES_FILE = existsSync(join(CAREER_OPS, 'templates/states.yml'))
  ? join(CAREER_OPS, 'templates/states.yml')
  : join(CAREER_OPS, 'states.yml');

// Ensure required directories exist (fresh setup)
mkdirSync(join(CAREER_OPS, 'data'), { recursive: true });
mkdirSync(REPORTS_DIR, { recursive: true });

const CANONICAL_STATUSES = [
  'evaluated', 'applied', 'responded', 'interview',
  'offer', 'rejected', 'discarded', 'skip',
];

const ALIASES = {
  'evaluada': 'evaluated', 'condicional': 'evaluated', 'hold': 'evaluated', 'evaluar': 'evaluated', 'verificar': 'evaluated',
  'aplicado': 'applied', 'enviada': 'applied', 'aplicada': 'applied', 'applied': 'applied', 'sent': 'applied',
  'respondido': 'responded',
  'entrevista': 'interview',
  'oferta': 'offer',
  'rechazado': 'rejected', 'rechazada': 'rejected',
  'descartado': 'discarded', 'descartada': 'discarded', 'cerrada': 'discarded', 'cancelada': 'discarded',
  'no aplicar': 'skip', 'no_aplicar': 'skip', 'monitor': 'skip', 'geo blocker': 'skip',
};

let errors = 0;
let warnings = 0;

function error(msg) { console.log(`❌ ${msg}`); errors++; }
function warn(msg) { console.log(`⚠️  ${msg}`); warnings++; }
function ok(msg) { console.log(`✅ ${msg}`); }

// --- Read applications.md ---
if (!existsSync(APPS_FILE)) {
  console.log('\n📊 No applications.md found. This is normal for a fresh setup.');
  console.log('   The file will be created when you evaluate your first offer.\n');
  process.exit(0);
}
const content = readFileSync(APPS_FILE, 'utf-8');
const lines = content.split('\n');

// Map columns by header name so the checks work whether the tracker uses the
// original 9-column layout or a customized one with an extra column (e.g. a
// Location column after Role). Fixed-position indexing would otherwise read
// Location where Score is expected and flag false errors. Falls back to the
// legacy fixed layout when no recognizable header row is found.
// Column detection is shared via column-map.mjs (resolveColumns falls back to
// the legacy fixed layout when no recognizable header row is found).
const COLMAP = resolveColumns(lines);
const MAX_IDX = Math.max(...Object.values(COLMAP));

const entries = [];
for (const line of lines) {
  if (!line.startsWith('|')) continue;
  const parts = line.split('|').map(s => s.trim());
  if (parts.length <= MAX_IDX) continue;
  const num = parseInt(parts[COLMAP.num]);
  if (isNaN(num)) continue;
  entries.push({
    num,
    date: parts[COLMAP.date],
    company: parts[COLMAP.company],
    role: parts[COLMAP.role],
    location: COLMAP.location != null ? parts[COLMAP.location] : '',
    score: parts[COLMAP.score],
    status: parts[COLMAP.status],
    pdf: parts[COLMAP.pdf],
    report: parts[COLMAP.report],
    notes: COLMAP.notes != null ? (parts[COLMAP.notes] || '') : '',
  });
}

console.log(`\n📊 Checking ${entries.length} entries in applications.md\n`);

// --- Check 1: Canonical statuses ---
let badStatuses = 0;
for (const e of entries) {
  const clean = e.status.replace(/\*\*/g, '').trim().toLowerCase();
  // Strip trailing dates
  const statusOnly = clean.replace(/\s+\d{4}-\d{2}-\d{2}.*$/, '').trim();

  if (!CANONICAL_STATUSES.includes(statusOnly) && !ALIASES[statusOnly]) {
    error(`#${e.num}: Non-canonical status "${e.status}"`);
    badStatuses++;
  }

  // Check for markdown bold in status
  if (e.status.includes('**')) {
    error(`#${e.num}: Status contains markdown bold: "${e.status}"`);
    badStatuses++;
  }

  // Check for dates in status
  if (/\d{4}-\d{2}-\d{2}/.test(e.status)) {
    error(`#${e.num}: Status contains date: "${e.status}" — dates go in date column`);
    badStatuses++;
  }
}
if (badStatuses === 0) ok('All statuses are canonical');

// --- Check 2: Duplicates ---
const companyRoleMap = new Map();
let dupes = 0;
for (const e of entries) {
  const key = e.company.toLowerCase().replace(/[^a-z0-9]/g, '') + '::' +
    e.role.toLowerCase().replace(/[^a-z0-9 ]/g, '');
  if (!companyRoleMap.has(key)) companyRoleMap.set(key, []);
  companyRoleMap.get(key).push(e);
}
for (const [key, group] of companyRoleMap) {
  if (group.length > 1) {
    warn(`Possible duplicates: ${group.map(e => `#${e.num}`).join(', ')} (${group[0].company} — ${group[0].role})`);
    dupes++;
  }
}
if (dupes === 0) ok('No exact duplicates found');

// --- Check 3: Report links ---
// Markdown links resolve relative to the file that contains them, so report
// links must resolve against the tracker's own directory (see #760). For the
// transition we also accept legacy root-relative links: try the tracker dir
// first, then fall back to the repo root before flagging a link broken.
const TRACKER_DIR = dirname(APPS_FILE);
let brokenReports = 0;
for (const e of entries) {
  const match = e.report.match(/\]\(([^)]+)\)/);
  if (!match) continue;
  const link = match[1];
  if (!existsSync(join(TRACKER_DIR, link)) && !existsSync(join(CAREER_OPS, link))) {
    error(`#${e.num}: Report not found: ${link}`);
    brokenReports++;
  }
}
if (brokenReports === 0) ok('All report links valid');

// --- Check 4: Score format ---
let badScores = 0;
for (const e of entries) {
  const s = e.score.replace(/\*\*/g, '').trim();
  if (!/^\d+\.?\d*\/5$/.test(s) && s !== 'N/A' && s !== 'DUP') {
    error(`#${e.num}: Invalid score format: "${e.score}"`);
    badScores++;
  }
}
if (badScores === 0) ok('All scores valid');

// --- Check 5: Row format ---
let badRows = 0;
for (const line of lines) {
  if (!line.startsWith('|')) continue;
  if (line.includes('---') || line.includes('Empresa')) continue;
  const parts = line.split('|');
  if (parts.length <= MAX_IDX) {
    error(`Row with too few columns (need ${MAX_IDX} data cols): ${line.substring(0, 80)}...`);
    badRows++;
  }
}
if (badRows === 0) ok('All rows properly formatted');

// --- Check 6: Pending TSVs ---
let pendingTsvs = 0;
if (existsSync(ADDITIONS_DIR)) {
  const files = readdirSync(ADDITIONS_DIR).filter(f => f.endsWith('.tsv'));
  pendingTsvs = files.length;
  if (pendingTsvs > 0) {
    warn(`${pendingTsvs} pending TSVs in tracker-additions/ (not merged)`);
  }
}
if (pendingTsvs === 0) ok('No pending TSVs');

// --- Check 7: Bold in scores ---
let boldScores = 0;
for (const e of entries) {
  if (e.score.includes('**')) {
    warn(`#${e.num}: Score has markdown bold: "${e.score}"`);
    boldScores++;
  }
}
if (boldScores === 0) ok('No bold in scores');

// --- Check 8: Stale report-number sentinels (GC) ---
// reserve-report-num.mjs drops NNN-RESERVED.md files in reports/ when a
// number is claimed.  If the process crashed before writing the real report
// and deleting the sentinel it will linger.  Sentinels older than 4 h are
// stale; remove them here so they don't skew the next slot allocation.
const SENTINEL_MAX_AGE_MS = 4 * 60 * 60 * 1000;
let staleSentinels = 0;
if (existsSync(REPORTS_DIR)) {
  const now = Date.now();
  for (const name of readdirSync(REPORTS_DIR)) {
    if (!name.endsWith('-RESERVED.md')) continue;
    const full = join(REPORTS_DIR, name);
    try {
      const { mtimeMs } = statSync(full);
      if (now - mtimeMs > SENTINEL_MAX_AGE_MS) {
        unlinkSync(full);
        warn(`Removed stale reservation sentinel: ${name}`);
        staleSentinels++;
      }
    } catch {
      // Already gone between readdir and stat — fine.
    }
  }
}
if (staleSentinels === 0) ok('No stale reservation sentinels');

// --- Check 9: Duplicate report file numbers ---
// reserve-report-num.mjs and batch runs can collide on a number and write two
// files sharing the same NNN- prefix. Earlier checks only catch duplicate
// tracker rows, so the orphaned twin stays invisible to any script that
// iterates reports by number.
let dupReportNums = 0;
if (existsSync(REPORTS_DIR)) {
  const numMap = new Map();
  for (const name of readdirSync(REPORTS_DIR)) {
    if (!name.endsWith('.md') || name.endsWith('-RESERVED.md')) continue;
    const m = name.match(/^(\d+)-/);
    if (!m) continue;
    if (!numMap.has(m[1])) numMap.set(m[1], []);
    numMap.get(m[1]).push(name);
  }
  for (const [n, files] of numMap) {
    if (files.length > 1) {
      warn(`Duplicate report number ${n}: ${files.join(', ')} — renumber the orphan to a free slot`);
      dupReportNums++;
    }
  }
}
if (dupReportNums === 0) ok('No duplicate report numbers');

// --- Check 10: Machine Summary present on actionable reports ---
// analyze-patterns.mjs reads a report's metadata from its `## Machine Summary`
// fenced block. Applied/Responded/Interview/Offer rows are the actionable ones
// whose metadata feeds pattern analysis — warn when their report lacks it.
const ACTIONABLE_STATUSES = new Set(['applied', 'responded', 'interview', 'offer']);
let missingMachineSummary = 0;
for (const e of entries) {
  const st = e.status.replace(/\*\*/g, '').trim().toLowerCase().replace(/\s+\d{4}-\d{2}-\d{2}.*$/, '').trim();
  const canon = CANONICAL_STATUSES.includes(st) ? st : (ALIASES[st] || st);
  if (!ACTIONABLE_STATUSES.has(canon)) continue;
  const match = e.report.match(/\]\(([^)]+)\)/);
  if (!match) continue;
  const link = match[1];
  const path = existsSync(join(TRACKER_DIR, link)) ? join(TRACKER_DIR, link)
    : existsSync(join(CAREER_OPS, link)) ? join(CAREER_OPS, link) : null;
  if (!path) continue; // broken link already flagged by the report-links check
  // Mirror analyze-patterns.mjs's parseMachineSummary regex: a bare heading with
  // no fenced YAML block still yields null there, so require the fence too.
  const MACHINE_SUMMARY_RE = /##\s*Machine Summary\s*\n+```(?:yaml|yml|json)?\s*\n[\s\S]*?\n```/i;
  if (!MACHINE_SUMMARY_RE.test(readFileSync(path, 'utf-8'))) {
    warn(`#${e.num} (${canon}): report missing a parseable "## Machine Summary" block — analyze-patterns can't read its metadata`);
    missingMachineSummary++;
  }
}
if (missingMachineSummary === 0) ok('All actionable reports have a Machine Summary');

// --- Summary ---
console.log('\n' + '='.repeat(50));
console.log(`📊 Pipeline Health: ${errors} errors, ${warnings} warnings`);
if (errors === 0 && warnings === 0) {
  console.log('🟢 Pipeline is clean!');
} else if (errors === 0) {
  console.log('🟡 Pipeline OK with warnings');
} else {
  console.log('🔴 Pipeline has errors — fix before proceeding');
}

process.exit(errors > 0 ? 1 : 0);

File: scripts/python/tracker/verify_pipeline.py
#!/usr/bin/env python3
from __future__ import annotations

import argparse
import re
from dataclasses import dataclass, field
from pathlib import Path

from scripts.python import PROJECT_ROOT, REPORTS_DIR, TEMPLATES_DIR
from scripts.python.tracker.parse import parse_tracker_row, resolve_columns
from scripts.python.tracker.utils import resolve_tracker_path


CANONICAL_STATUSES = {"evaluated", "applied", "responded", "interview", "offer", "rejected", "discarded", "skip", "hired"}
ALIASES = {
    "evaluada": "evaluated",
    "condicional": "evaluated",
    "hold": "evaluated",
    "evaluar": "evaluated",
    "verificar": "evaluated",
    "aplicado": "applied",
    "enviada": "applied",
    "aplicada": "applied",
    "applied": "applied",
    "sent": "applied",
    "respondido": "responded",
    "entrevista": "interview",
    "oferta": "offer",
    "rechazado": "rejected",
    "rechazada": "rejected",
    "descartado": "discarded",
    "descartada": "discarded",
    "cerrada": "discarded",
    "cancelada": "discarded",
    "no aplicar": "skip",
    "no_aplicar": "skip",
    "monitor": "skip",
    "geo blocker": "skip",
    "contratado": "hired",
    "contratada": "hired",
    "hired": "hired",
    "accepted": "hired",
    "accept": "hired",
}
REPORT_FILE_RE = re.compile(r"^(\d+)-(.+)-\d{4}-\d{2}-\d{2}\.md$")


@dataclass
class VerifyResult:
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    @property
    def ok(self) -> bool:
        return not self.errors


def _normalize_key(value: str) -> str:
    return re.sub(r"[^a-z0-9]", "", str(value or "").lower())


def _extract_report_role(content: str) -> str | None:
    fence = re.search(r"##\s*Machine Summary\s*\n+```(?:yaml|yml|json)?\s*\n([\s\S]*?)\n```", content, re.IGNORECASE)
    if fence:
        match = re.search(r"^role:\s*[\"']?(.+?)[\"']?\s*$", fence.group(1), re.MULTILINE)
        if match and match.group(1).strip():
            return match.group(1).strip()
    title = next((line for line in content.splitlines() if line.startswith("# ")), "")
    parts = re.split(r"[—–]", title)
    return parts[-1].strip() if len(parts) >= 2 and parts[-1].strip() else None


def verify_pipeline(
    tracker_path: str | Path | None = None,
    *,
    reports_dir: str | Path | None = None,
    additions_dir: str | Path | None = None,
) -> VerifyResult:
    apps_file = Path(tracker_path) if tracker_path else resolve_tracker_path()
    reports_path = Path(reports_dir or REPORTS_DIR)
    additions_path = Path(additions_dir or PROJECT_ROOT / "batch/tracker-additions")
    result = VerifyResult()
    if not apps_file.exists():
        return result

    content = apps_file.read_text(encoding="utf-8")
    lines = content.split("\n")
    colmap = resolve_columns(lines)
    max_idx = max(colmap.values())
    rows = [row for line in lines if (row := parse_tracker_row(line, colmap))]

    for row in rows:
        status = re.sub(r"\s+\d{4}-\d{2}-\d{2}.*$", "", row.status.replace("**", "").strip().lower())
        if status not in CANONICAL_STATUSES and status not in ALIASES:
            result.errors.append(f"#{row.num}: Non-canonical status \"{row.status}\"")
        if "**" in row.status:
            result.errors.append(f"#{row.num}: Status contains markdown bold")
        if re.search(r"\d{4}-\d{2}-\d{2}", row.status):
            result.errors.append(f"#{row.num}: Status contains date")

        score = row.score.replace("**", "").strip()
        if not re.match(r"^\d+\.?\d*/5$", score) and score not in {"N/A", "DUP"}:
            result.errors.append(f"#{row.num}: Invalid score format \"{row.score}\"")
        if "**" in row.score:
            result.warnings.append(f"#{row.num}: Score has markdown bold")

        link_match = re.search(r"\]\(([^)]+)\)", row.report)
        if link_match:
            link = link_match.group(1)
            if not (apps_file.parent / link).exists() and not (PROJECT_ROOT / link).exists():
                result.errors.append(f"#{row.num}: Report not found: {link}")

    seen: dict[str, list[object]] = {}
    for row in rows:
        key = f"{_normalize_key(row.company)}::{re.sub(r'[^a-z0-9 ]', '', row.role.lower())}"
        seen.setdefault(key, []).append(row)
    for group in seen.values():
        if len(group) > 1:
            result.warnings.append(f"Possible duplicates: {', '.join(f'#{row.num}' for row in group)}")

    for line in lines:
        if not line.startswith("|") or "---" in line or "Empresa" in line:
            continue
        if len(line.split("|")) <= max_idx:
            result.errors.append(f"Row with too few columns: {line[:80]}")

    if additions_path.exists():
        pending = [path for path in additions_path.iterdir() if path.suffix == ".tsv"]
        if pending:
            result.warnings.append(f"{len(pending)} pending TSVs in tracker-additions/")

    report_files = [path for path in reports_path.iterdir() if REPORT_FILE_RE.match(path.name)] if reports_path.exists() else []
    reports_by_role: dict[str, list[str]] = {}
    for path in report_files:
        match = REPORT_FILE_RE.match(path.name)
        role = _extract_report_role(path.read_text(encoding="utf-8")) if path.exists() else None
        if not match or not role:
            continue
        reports_by_role.setdefault(f"{_normalize_key(match.group(2))}::{_normalize_key(role)}", []).append(path.name)
    for group in reports_by_role.values():
        if len(group) > 1:
            result.warnings.append(f"Duplicate reports for same company+role: {', '.join(group)}")

    referenced_nums: set[int] = set()
    for row in rows:
        referenced_nums.add(row.num)
        link_text = re.search(r"\[(\d+)\]", row.report)
        if link_text:
            referenced_nums.add(int(link_text.group(1)))
        link_target = re.search(r"\]\(([^)]+)\)", row.report)
        if link_target:
            target_name = link_target.group(1).split("/")[-1]
            prefix = re.match(r"^(\d+)-", target_name)
            if prefix:
                referenced_nums.add(int(prefix.group(1)))
    for path in report_files:
        num = int(REPORT_FILE_RE.match(path.name).group(1))  # type: ignore[union-attr]
        if num not in referenced_nums:
            result.warnings.append(f"Orphan report: reports/{path.name}")

    nums: dict[int, list[object]] = {}
    for row in rows:
        nums.setdefault(row.num, []).append(row)
    for num, group in nums.items():
        if len(group) > 1:
            result.errors.append(f"Duplicate tracker number #{num}")

    if TEMPLATES_DIR.joinpath("states.yml").exists():
        text = TEMPLATES_DIR.joinpath("states.yml").read_text(encoding="utf-8")
        for state in CANONICAL_STATUSES - {"hired"}:
            if f"id: {state}" not in text:
                result.warnings.append(f"states.yml missing canonical id: {state}")
    return result


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(description="Check career-ops pipeline integrity.")
    parser.add_argument("--tracker")
    parser.add_argument("--reports")
    parser.add_argument("--additions")
    return parser


def main(argv: list[str] | None = None) -> int:
    args = build_parser().parse_args(argv)
    result = verify_pipeline(args.tracker, reports_dir=args.reports, additions_dir=args.additions)
    for message in result.errors:
        print(f"ERROR: {message}")
    for message in result.warnings:
        print(f"WARN: {message}")
    print(f"Pipeline Health: {len(result.errors)} errors, {len(result.warnings)} warnings")
    return 1 if result.errors else 0


if __name__ == "__main__":
    raise SystemExit(main())
